nautobot/docs/development/apps/migration/ui-component-framework.md

```diff
@@ -0,0 +1,316 @@
+from nautobot.apps.ui import KeyValueTablePanel
+from nautobot.core.templatetags.helpers import (
+    divide,
+    placeholder,
+    slugify,
+    split,
+)
+
+class DeviceDetailView(views.NautobotUIViewSet):
+    panels = [
+        KeyValueTablePanel(weight=100, value_transforms={
+            "name": [slugify, placeholder],
+            "list_of_names": split,
+            "number_value": lambda v: divide(v, 3),
+        }),
+    ]
+
+    def get_extra_context(self, request, instance):
+        context = super().get_extra_context(request, instance)
+        context["data"] = {
+            "name": "Some Example Name",
+            "list_of_names": "Name1 Name2 Name3",
+            "number_value": 1000,
+        }
+        return context
+```
+
+More built-in filters can be found in [`Nautobot Built-In Filters`](../../../user-guide/platform-functionality/template-filters.md#nautobot-built-in-filters)
+
+
+3. Create a custom Panel class if needed:
+
+```python title="custom_panel.py"
+from django.template.loader import get_template
+from nautobot.apps.ui import Panel
+
+class CustomPanel(Panel):
+
+    # You can override default body panel
+    def __init__(self, *, body_content_template_path="custom_template_path.html", **kwargs):
+        super().__init__(body_content_template_path=body_content_template_path, **kwargs)
+
+    # Adding some extra context specially for this panel / or this custom template
+    def get_extra_context(self, context):
+        # This context will be passed to the label, header, body and footer render methods
+        return {"custom_data": "I love Nautobot!"}
+```
+```python title="views.py"
+from nautobot.apps import views
+from device_app.custom_panel import CustomPanel
+
+class DeviceDetailView(views.NautobotUIViewSet):
+    panels = [
+        CustomPanel(weight=100),
+    ]
+```
+
+If need more custom behaviour, you can override other `Panel` rendering methods. For more details please refer to the [`Panel` Code Reference.](#todo: link goes here)
+
+## Best Practices for Migration
+
+1. **Incremental Migration**
+   - Migrate one view at a time
+   - Test thoroughly after each conversion
+   - Keep backup of template-based views
+
+2. **Panel Organization**
+   - Use weights relative to the `Panel.`WEIGHT_*_PANEL` constants for consistency.
+   - Group related panels together
+
+3. **Performance Considerations**
+   - Use select_related/prefetch_related in ObjectsTablePanel (though note that BaseTable may handle some simple optimizations automatically)
+   - Optimize StatsPanel queries
+   - Cache complex transformations
+
+4. **Common Patterns**
+```python
+# Statistics and related objects
+StatsPanel(
+    filter_name="device_type",
+    related_models=[Device],
+    section=SectionChoices.RIGHT_HALF,
+    weight=100,
+)
+
+# Custom fields grouping
+GroupedKeyValueTablePanel(
+    body_id="custom-fields",
+    data=self.get_custom_field_groups(),
+    section=SectionChoices.FULL_WIDTH,
+    weight=200,
+)
```

Comment:
Probably remove this example, or replace with `ObjectFieldsPanel` and `ObjectsTablePanel` examples?